File: apps/api/src/composio.ts
import type { D1Like } from "@rel/core";
import type { Env } from "./env";

/**
 * Composio, from the Worker.
 *
 * The connect flow is the one proven in the AIOS dashboard
 * (`wt-core-pr37-audit/src/connectors/composio.ts`): one managed auth config per
 * toolkit, then `POST /connected_accounts/link` for the user to sign in — with
 * the raw `/connected_accounts` call kept as a fallback while Composio retires
 * it. Tool execution is the same three-part contract our Python connector uses
 * (toolkit version, an ACTIVE account, and that account's own user id).
 *
 * Nothing here pretends a connection exists: an unfinished sign-in stays
 * pending, and a missing key says so instead of failing obscurely.
 */

const BASE = "https://backend.composio.dev/api/v3";

export class ComposioError extends Error {
  status: number;
  constructor(message: string, status = 502) {
    super(message);
    this.status = status;
  }
}

export interface ComposioAccount {
  id: string;
  toolkit: string;
  status: string;
  userId: string;
  label: string;
  alias: string | null;
}

/** What the graph can actually read, and how it says so on screen. */
export interface ToolkitSpec {
  slug: string;
  source: string;
  label: string;
  reads: string;
}

export const TOOLKITS: ToolkitSpec[] = [
  {
    slug: "gmail",
    source: "gmail",
    label: "Gmail",
    reads: "Messages and threads — who writes to you, and who you actually answer.",
  },
  {
    slug: "googlecalendar",
    source: "google-calendar",
    label: "Google Calendar",
    reads: "Events and attendees — the people you meet with, not just the ones who email.",
  },
  {
    slug: "fathom",
    source: "fathom",
    label: "Fathom",
    reads: "Recorded calls with attendees, summaries and action items.",
  },
];

export function toolkitFor(value: string): ToolkitSpec | null {
  const wanted = value.trim().toLowerCase();
  return TOOLKITS.find((toolkit) => toolkit.slug === wanted || toolkit.source === wanted) ?? null;
}

export function composioConfigured(env: Env): boolean {
  return Boolean(env.COMPOSIO_API_KEY && env.COMPOSIO_API_KEY.trim());
}

async function composioFetch(env: Env, path: string, init: RequestInit = {}): Promise<Response> {
  const key = env.COMPOSIO_API_KEY?.trim();
  if (!key) throw new ComposioError("COMPOSIO_API_KEY is not set on this deployment", 503);

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 25_000);
  try {
    return await fetch(`${BASE}${path}`, {
      ...init,
      signal: controller.signal,
      headers: { "x-api-key": key, "Content-Type": "application/json", ...(init.headers ?? {}) },
    });
  } catch (error) {
    throw new ComposioError(
      `Composio could not be reached (${error instanceof Error ? error.message : String(error)})`,
      502,
    );
  } finally {
    clearTimeout(timer);
  }
}

function asRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" ? (value as Record<string, unknown>) : {};
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

async function composioJson(env: Env, path: string, init: RequestInit = {}): Promise<Record<string, unknown>> {
  const response = await composioFetch(env, path, init);
  const text = await response.text();
  if (!response.ok) {
    if (response.status === 401 || response.status === 403) {
      throw new ComposioError("Composio rejected the API key", 502);
    }
    throw new ComposioError(
      `Composio ${path} failed (${response.status})${text ? `: ${text.slice(0, 240)}` : ""}`,
      502,
    );
  }
  if (!text) return {};
  try {
    return asRecord(JSON.parse(text));
  } catch {
    throw new ComposioError(`Composio ${path} returned something that is not JSON`, 502);
  }
}

/* ---- auth configs ---- */

/**
 * One managed auth config per toolkit, created on first connect and remembered,
 * so a second account is a sign-in and not another config.
 */
export async function ensureAuthConfig(env: Env, db: D1Like, toolkit: string): Promise<string> {
  const stored = await db
    .prepare("SELECT auth_config_id FROM toolkit_auth_configs WHERE toolkit = ?")
    .bind(toolkit)
    .first<{ auth_config_id: string }>();
  if (stored?.auth_config_id) return stored.auth_config_id;

  const query = new URLSearchParams({ toolkit_slug: toolkit, is_composio_managed: "true" });
  const existing = await composioJson(env, `/auth_configs?${query.toString()}`);
  const first = asArray(existing.items)[0];
  let authConfigId = String(asRecord(first).id ?? asRecord(first).nanoid ?? "");

  if (!authConfigId) {
    const created = await composioJson(env, "/auth_configs", {
      method: "POST",
      body: JSON.stringify({
        toolkit: { slug: toolkit },
        auth_config: { type: "use_composio_managed_auth", name: `relationship-manager-${toolkit}` },
      }),
    }).catch((error: unknown) => {
      if (error instanceof ComposioError && error.status === 502) {
        throw new ComposioError(
          `${toolkit} needs its own OAuth developer app before it can be connected here — Composio refused to create a managed auth config (${error.message.slice(0, 160)})`,
          400,
        );
      }
      throw error;
    });
    authConfigId = String(asRecord(created.auth_config).id ?? created.id ?? created.nanoid ?? "");
  }

  if (!authConfigId) throw new ComposioError(`Composio returned no auth config for ${toolkit}`, 502);

  await db
    .prepare(
      `INSERT INTO toolkit_auth_configs (toolkit, auth_config_id, managed, created_at) VALUES (?, ?, 1, ?)
       ON CONFLICT(toolkit) DO UPDATE SET auth_config_id = excluded.auth_config_id`,
    )
    .bind(toolkit, authConfigId, new Date().toISOString())
    .run();

  return authConfigId;
}

/* ---- accounts ---- */

function accountFrom(value: unknown): ComposioAccount | null {
  const item = asRecord(value);
  const id = String(item.id ?? item.nanoid ?? "");
  if (!id) return null;
  const toolkit = String(asRecord(item.toolkit).slug ?? item.toolkit ?? "");
  const label = String(
    item.account_label ??
      item.accountLabel ??
      item.account_email ??
      item.accountEmail ??
      item.email ??
      asRecord(item.params).email ??
      asRecord(item.metadata).email ??
      item.word_id ??
      item.account_id ??
      "",
  );
  return {
    id,
    toolkit,
    status: String(item.status ?? item.state ?? "PENDING").toUpperCase(),
    userId: String(item.user_id ?? item.userId ?? ""),
    label,
    alias: typeof item.alias === "string" ? item.alias : null,
  };
}

export async function listAccounts(env: Env): Promise<ComposioAccount[]> {
  const body = await composioJson(env, "/connected_accounts?limit=100");
  return asArray(body.items).map(accountFrom).filter((account): account is ComposioAccount => account !== null);
}

export async function getAccount(env: Env, id: string): Promise<ComposioAccount | null> {
  const body = await composioJson(env, `/connected_accounts/${encodeURIComponent(id)}`);
  return accountFrom(body);
}

export async function removeAccount(env: Env, id: string): Promise<void> {
  const response = await composioFetch(env, `/connected_accounts/${encodeURIComponent(id)}`, { method: "DELETE" });
  // Already gone is the outcome we wanted.
  if (!response.ok && response.status !== 404) {
    throw new ComposioError(`Composio would not remove the account (${response.status})`, 502);
  }
}

/**
 * Start a sign-in. The modern link endpoint first; the raw create call is kept
 * as a fallback while Composio retires it (both AIOS implementations do this).
 */
export async function createConnectLink(
  env: Env,
  db: D1Like,
  toolkit: string,
  userId: string,
  callbackUrl?: string,
): Promise<{ connectionId: string; redirectUrl: string | null; active: boolean }> {
  const authConfigId = await ensureAuthConfig(env, db, toolkit);

  const link = await composioFetch(env, "/connected_accounts/link", {
    method: "POST",
    body: JSON.stringify({
      auth_config_id: authConfigId,
      user_id: userId,
      ...(callbackUrl ? { callback_url: callbackUrl } : {}),
    }),
  });

  if (link.ok) {
    const body = asRecord(await link.json().catch(() => ({})));
    const connectionId = String(body.connected_account_id ?? body.id ?? "");
    const redirectUrl = body.redirect_url ? String(body.redirect_url) : null;
    if (connectionId && (redirectUrl || body.status)) {
      const status = String(body.status ?? "").toUpperCase();
      return { connectionId, redirectUrl, active: status === "ACTIVE" || status === "CONNECTED" };
    }
  }

  // Fallback: the raw endpoint the SDK used to wrap.
  const created = await composioJson(env, "/connected_accounts", {
    method: "POST",
    body: JSON.stringify({ auth_config: { id: authConfigId }, connection: { user_id: userId } }),
  });
  const connectionId = String(created.id ?? created.connectedAccountId ?? created.connection_id ?? "");
  const redirectUrl = created.redirect_url ?? created.redirectUrl ?? created.authUrl ?? null;
  const status = String(created.status ?? "").toUpperCase();
  if (!connectionId) throw new ComposioError("Composio started no connection for this sign-in", 502);
  return { connectionId, redirectUrl: redirectUrl ? String(redirectUrl) : null, active: status === "ACTIVE" };
}

/* ---- tool execution ---- */

const versionCache = new Map<string, string>();

async function toolkitVersion(env: Env, toolkit: string): Promise<string> {
  const cached = versionCache.get(toolkit);
  if (cached) return cached;
  const body = await composioJson(env, `/toolkits/${encodeURIComponent(toolkit)}`);
  const versions = asArray(asRecord(body.meta).available_versions);
  const version = versions.length > 0 ? String(versions[0]) : "";
  if (!version) throw new ComposioError(`Composio exposes no version for ${toolkit}`, 502);
  versionCache.set(toolkit, version);
  return version;
}

export async function execute(
  env: Env,
  options: {
    slug: string;
    toolkit: string;
    account: ComposioAccount;
    arguments: Record<string, unknown>;
  },
): Promise<Record<string, unknown>> {
  const version = await toolkitVersion(env, options.toolkit);
  const body = await composioJson(env, `/tools/execute/${options.slug}`, {
    method: "POST",
    body: JSON.stringify({
      user_id: options.account.userId,
      connected_account_id: options.account.id,
      version,
      arguments: options.arguments,
    }),
  });
  const data = body.data;
  if (data === undefined || data === null) {
    throw new ComposioError(`${options.slug} returned no data`, 502);
  }
  return asRecord(data);
}

/* ---- the pulls (same shapes the Python connector pushes) ---- */

export async function fetchGmail(
  env: Env,
  account: ComposioAccount,
  query: string,
  maxResults = 100,
): Promise<Array<Record<string, unknown>>> {
  const data = await execute(env, {
    slug: "GMAIL_FETCH_EMAILS",
    toolkit: "gmail",
    account,
    arguments: { query, max_results: maxResults },
  });

  const rows = asArray(data.messages).length > 0 ? asArray(data.messages) : asArray(data.items);
  const messages: Array<Record<string, unknown>> = [];
  for (const value of rows) {
    const row = asRecord(value);
    const id = String(row.messageId ?? row.id ?? row.threadId ?? "");
    if (!id) continue;
    const labels = Array.isArray(row.labelIds) ? row.labelIds : Array.isArray(row.labels) ? row.labels : [];
    messages.push({
      id,
      service: "gmail",
      thread_id: row.threadId ?? id,
      from_addr: row.sender ?? row.from ?? "",
      to_addr: row.to ?? row.recipient ?? "",
      subject: row.subject ?? "",
      snippet: String(row.snippet ?? row.preview ?? "").slice(0, 1000),
      body_text: String(row.messageText ?? row.body ?? "").slice(0, 20000),
      labels: labels.map(String).join(","),
      is_unread: labels.includes("UNREAD") ? 1 : 0,
      last_from_user: labels.includes("SENT") ? 1 : 0,
      internal_date: row.messageTimestamp ?? row.internalDate ?? row.date ?? "",
    });
  }
  return messages;
}

export async function fetchCalendar(env: Env, account: ComposioAccount, daysBack = 30, daysForward = 30): Promise<Array<Record<string, unknown>>> {
  const now = Date.now();
  const data = await execute(env, {
    slug: "GOOGLECALENDAR_EVENTS_LIST",
    toolkit: "googlecalendar",
    account,
    arguments: {
      timeMin: new Date(now - daysBack * 86_400_000).toISOString(),
      timeMax: new Date(now + daysForward * 86_400_000).toISOString(),
      maxResults: 250,
    },
  });

  const rows = asArray(data.items).length > 0 ? asArray(data.items) : asArray(data.events);
  const events: Array<Record<string, unknown>> = [];
  for (const value of rows) {
    const row = asRecord(value);
    const id = String(row.id ?? "");
    if (!id) continue;
    const start = asRecord(row.start);
    const end = asRecord(row.end);
    const organizer = asRecord(row.organizer);
    events.push({
      id,
      service: "google-calendar",
      title: row.summary ?? "",
      start_at: start.dateTime ?? start.date ?? null,
      end_at: end.dateTime ?? end.date ?? null,
      is_all_day: start.date ? 1 : 0,
      location: row.location ?? "",
      organizer: organizer.email ?? "",
      attendees: asArray(row.attendees)
        .map((attendee) => String(asRecord(attendee).email ?? ""))
        .filter(Boolean)
        .join(","),
      status: row.status ?? "",
      link: row.htmlLink ?? "",
    });
  }
  return events;
}

export async function fetchFathom(
  env: Env,
  account: ComposioAccount,
  days = 120,
  maxPages = 6,
): Promise<Array<Record<string, unknown>>> {
  const createdAfter = new Date(Date.now() - days * 86_400_000).toISOString().replace(/\.\d{3}Z$/, "Z");
  const meetings: Array<Record<string, unknown>> = [];
  let cursor = "";

  for (let page = 0; page < maxPages; page += 1) {
    const arguments_: Record<string, unknown> = { created_after: createdAfter, include_action_items: true };
    if (cursor) arguments_.cursor = cursor;
    const data = await execute(env, { slug: "FATHOM_LIST_MEETINGS", toolkit: "fathom", account, arguments: arguments_ });

    for (const value of asArray(data.items)) {
      const item = asRecord(value);
      const recordingId = item.recording_id;
      if (!recordingId) continue;

      const attendees: string[] = [];
      for (const invitee of asArray(item.calendar_invitees)) {
        const person = asRecord(invitee);
        const handle = person.email ?? person.name;
        if (handle) attendees.push(String(handle));
      }
      const recorder = asRecord(item.recorded_by);
      if (recorder.email) attendees.push(String(recorder.email));

      meetings.push({
        id: `fathom_${recordingId}`,
        title: String(item.title ?? item.meeting_title ?? "(untitled meeting)").slice(0, 300),
        endedAt: item.recording_end_time ?? item.created_at,
        attendees: attendees.slice(0, 12),
        summary: String(item.default_summary ?? "").slice(0, 4000),
        actionItems: asArray(item.action_items)
          .map((action) => String(asRecord(action).description ?? action))
          .slice(0, 20),
        url: item.share_url ?? item.url ?? "",
      });
    }

    cursor = String(data.next_cursor ?? "");
    if (!cursor) break;
  }
  return meetings;
}

File: connectors/rel_sync/providers/composio.py
"""Composio — the live sources, over Composio's HTTP API.

Uses the API directly rather than the CLI: the CLI on this machine points at a
host that rejects the key, while the API accepts it. Tool execution needs three
things, and getting any of them wrong returns a bare "tool not found", so they
are resolved from the API rather than hard-coded:

  1. the toolkit's version  (`meta.available_versions[0]`)
  2. a connected account id that is ACTIVE
  3. the user id that account belongs to

What Composio has connected decides what can be synced. Nothing here pretends a
source is live when its account is missing or expired — `probe()` says which.
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Dict, List, Optional

from ..config import Config

BASE = "https://backend.composio.dev/api/v3"

# Which toolkits this connector knows how to read.
WANTED = {
    "fathom": "Fathom calls",
    "gmail": "Gmail",
    "googlecalendar": "Google Calendar",
}


class ComposioError(RuntimeError):
    pass


class Composio:
    def __init__(self, api_key: str) -> None:
        if not api_key:
            raise ComposioError("COMPOSIO_API_KEY is not set")
        self.api_key = api_key

    def _call(self, path: str, method: str = "GET", body: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        request = urllib.request.Request(
            "%s%s" % (BASE, path),
            data=json.dumps(body).encode() if body is not None else None,
            headers={"x-api-key": self.api_key, "Content-Type": "application/json"},
            method=method,
        )
        try:
            with urllib.request.urlopen(request, timeout=90) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as error:
            detail = error.read().decode("utf-8", "replace")
            if error.code in (401, 403):
                raise ComposioError("Composio rejected the key (%s)" % error.code) from error
            raise ComposioError("Composio %s failed (%s): %s" % (path, error.code, detail[:200])) from error
        except urllib.error.URLError as error:
            raise ComposioError("could not reach Composio: %s" % error.reason) from error

    # ---- discovery --------------------------------------------------------

    def accounts(self, toolkit: Optional[str] = None, active_only: bool = True) -> List[Dict[str, Any]]:
        payload = self._call("/connected_accounts?limit=100")
        accounts = payload.get("items", [])
        out = []
        for account in accounts:
            slug = (account.get("toolkit") or {}).get("slug", "")
            if toolkit and slug != toolkit:
                continue
            if active_only and account.get("status") != "ACTIVE":
                continue
            out.append(account)
        return out

    def versions(self, toolkit: str) -> List[str]:
        payload = self._call("/toolkits/%s" % toolkit)
        return (payload.get("meta") or {}).get("available_versions") or []

    def connected_toolkits(self) -> Dict[str, int]:
        counts: Dict[str, int] = {}
        for account in self._call("/connected_accounts?limit=100").get("items", []):
            slug = (account.get("toolkit") or {}).get("slug", "?")
            if account.get("status") == "ACTIVE":
                counts[slug] = counts.get(slug, 0) + 1
        return counts

    def execute(
        self,
        slug: str,
        arguments: Dict[str, Any],
        toolkit: str,
        account: Optional[Dict[str, Any]] = None,
        version: Optional[str] = None,
    ) -> Dict[str, Any]:
        account = account or (self.accounts(toolkit) or [None])[0]
        if not account:
            raise ComposioError(
                "No ACTIVE %s account is connected in Composio. Connect one, then re-run." % toolkit
            )
        versions = self.versions(toolkit)
        resolved_version = version or (versions[0] if versions else None)
        if not resolved_version:
            raise ComposioError("Composio exposes no version for toolkit %s" % toolkit)

        payload = self._call(
            "/tools/execute/%s" % slug,
            method="POST",
            body={
                "user_id": account.get("user_id"),
                "connected_account_id": account["id"],
                "version": resolved_version,
                "arguments": arguments,
            },
        )
        data = payload.get("data")
        if data is None:
            raise ComposioError("%s returned no data: %s" % (slug, json.dumps(payload)[:200]))
        return data


# ---------------------------------------------------------------------------
# Fathom
# ---------------------------------------------------------------------------


def fetch_fathom_meetings(config: Config, days: int = 120, max_pages: int = 12) -> List[Dict[str, Any]]:
    """Recorded calls. Summaries need an API-key connection, so on OAuth we take
    what Fathom will give: who, what, when, and the links."""
    import datetime

    client = Composio(config.composio_api_key)
    created_after = (
        datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(days=days)
    ).strftime("%Y-%m-%dT%H:%M:%SZ")

    meetings: List[Dict[str, Any]] = []
    cursor: Optional[str] = None
    for _ in range(max_pages):
        arguments: Dict[str, Any] = {"created_after": created_after, "include_action_items": True}
        if cursor:
            arguments["cursor"] = cursor
        data = client.execute("FATHOM_LIST_MEETINGS", arguments, "fathom")
        items = data.get("items") or []
        for item in items:
            recording_id = item.get("recording_id")
            if not recording_id:
                continue
            attendees: List[str] = []
            for invitee in item.get("calendar_invitees") or []:
                if isinstance(invitee, dict):
                    value = invitee.get("email") or invitee.get("name")
                    if value:
                        attendees.append(str(value))
            recorder = item.get("recorded_by")
            if isinstance(recorder, dict) and recorder.get("email"):
                attendees.append(str(recorder["email"]))

            meetings.append(
                {
                    "id": "fathom_%s" % recording_id,
                    "title": str(item.get("title") or item.get("meeting_title") or "(untitled meeting)")[:300],
                    "endedAt": item.get("recording_end_time") or item.get("created_at"),
                    "attendees": attendees[:12],
                    "summary": str(item.get("default_summary") or "")[:4000],
                    "actionItems": [
                        str(action.get("description") if isinstance(action, dict) else action)
                        for action in (item.get("action_items") or [])
                    ][:20],
                    "url": item.get("share_url") or item.get("url") or "",
                }
            )
        cursor = data.get("next_cursor") or ""
        if not cursor:
            break
    return meetings


# ---------------------------------------------------------------------------
# Gmail and Calendar (only when those toolkits are connected)
# ---------------------------------------------------------------------------

GMAIL_TOOL = "GMAIL_FETCH_EMAILS"
CALENDAR_TOOL = "GOOGLECALENDAR_EVENTS_LIST"


def fetch_gmail(config: Config, query: str = "newer_than:3d", max_results: int = 50) -> List[Dict[str, Any]]:
    client = Composio(config.composio_api_key)
    data = client.execute(GMAIL_TOOL, {"query": query, "max_results": max_results}, "gmail")
    rows = data.get("messages") or data.get("items") or []
    messages: List[Dict[str, Any]] = []
    for row in rows:
        if not isinstance(row, dict):
            continue
        message_id = str(row.get("messageId") or row.get("id") or row.get("threadId") or "")
        if not message_id:
            continue
        labels = row.get("labelIds") or row.get("labels") or []
        messages.append(
            {
                "id": message_id,
                "service": "gmail",
                "thread_id": row.get("threadId") or message_id,
                "from_addr": row.get("sender") or row.get("from") or "",
                "to_addr": row.get("to") or row.get("recipient") or "",
                "subject": row.get("subject") or "",
                "snippet": str(row.get("snippet") or row.get("preview") or "")[:1000],
                "body_text": str(row.get("messageText") or row.get("body") or "")[:20000],
                "labels": ",".join(labels) if isinstance(labels, list) else str(labels or ""),
                "is_unread": 1 if "UNREAD" in labels else 0,
                "last_from_user": 1 if "SENT" in labels else 0,
                "internal_date": row.get("messageTimestamp") or row.get("internalDate") or row.get("date") or "",
            }
        )
    return messages


def fetch_calendar(config: Config, days_back: int = 14, days_forward: int = 14) -> List[Dict[str, Any]]:
    import datetime

    client = Composio(config.composio_api_key)
    now = datetime.datetime.now(datetime.timezone.utc)
    data = client.execute(
        CALENDAR_TOOL,
        {
            "timeMin": (now - datetime.timedelta(days=days_back)).isoformat(),
            "timeMax": (now + datetime.timedelta(days=days_forward)).isoformat(),
            "maxResults": 250,
        },
        "googlecalendar",
    )
    rows = data.get("items") or data.get("events") or []
    events: List[Dict[str, Any]] = []
    for row in rows:
        if not isinstance(row, dict) or not row.get("id"):
            continue
        start = row.get("start") or {}
        end = row.get("end") or {}
        events.append(
            {
                "id": str(row["id"]),
                "service": "google-calendar",
                "title": row.get("summary") or "",
                "start_at": (start.get("dateTime") or start.get("date")) if isinstance(start, dict) else start,
                "end_at": (end.get("dateTime") or end.get("date")) if isinstance(end, dict) else end,
                "is_all_day": 1 if isinstance(start, dict) and start.get("date") else 0,
                "location": row.get("location") or "",
                "organizer": (row.get("organizer") or {}).get("email", "") if isinstance(row.get("organizer"), dict) else "",
                "attendees": ",".join(
                    str(a.get("email", "")) for a in (row.get("attendees") or []) if isinstance(a, dict)
                ),
                "status": row.get("status") or "",
                "link": row.get("htmlLink") or "",
            }
        )
    return events


def probe(config: Config) -> Optional[str]:
    """None when Composio is usable, otherwise one plain sentence saying why not."""
    if not config.composio_api_key:
        return "COMPOSIO_API_KEY is not set"
    try:
        toolkits = Composio(config.composio_api_key).connected_toolkits()
    except ComposioError as error:
        return str(error)
    if not toolkits:
        return "Composio has no connected accounts"
    wanted = ", ".join("%s×%d" % (slug, count) for slug, count in sorted(toolkits.items()) if slug in WANTED)
    return None if wanted else "nothing this connector can read is connected (found: %s)" % ", ".join(sorted(toolkits))
